src/app/wallets/reimburse-fee.ts

```diff
@@ -1,89 +1,109 @@
+import { toCents } from "@domain/fiat"
+import { LedgerTransactionType } from "@domain/ledger"
 import { FeeReimbursement } from "@domain/ledger/fee-reimbursement"
-import { WalletCurrency } from "@domain/shared"
-import { DealerPriceService } from "@services/dealer-price"
-import { LedgerService } from "@services/ledger"
+import { PriceRatio } from "@domain/payments"
+import { paymentAmountFromSats, WalletCurrency } from "@domain/shared"
 
-export const reimburseFee = async ({
-  walletId,
-  walletCurrency,
+import * as LedgerFacade from "@services/ledger/facade"
+
+export const reimburseFee = async <S extends WalletCurrency, R extends WalletCurrency>({
+  paymentFlow,
   journalId,
-  paymentHash,
-  maxFee,
   actualFee,
   revealedPreImage,
-  paymentAmount,
-  usdFee,
+  amountDisplayCurrency,
+  feeDisplayCurrency,
+  displayCurrency,
   logger,
 }: {
-  walletId: WalletId
-  walletCurrency: WalletCurrency
+  paymentFlow: PaymentFlow<S, R>
   journalId: LedgerJournalId
-  paymentHash: PaymentHash
-  maxFee: Satoshis
   actualFee: Satoshis
   revealedPreImage?: RevealedPreImage
-  paymentAmount: Satoshis
-  usdFee: DisplayCurrencyBaseAmount
+  amountDisplayCurrency: DisplayCurrencyBaseAmount
+  feeDisplayCurrency: DisplayCurrencyBaseAmount
+  displayCurrency: DisplayCurrency
   logger: Logger
 }): Promise<true | ApplicationError> => {
-  let cents: UsdCents | undefined
+  const actualFeeAmount = paymentAmountFromSats(actualFee)
+
+  const maxFeeAmounts = {
+    btc: paymentFlow.btcProtocolFee,
+    usd: paymentFlow.usdProtocolFee,
+  }
 
-  const feeDifference = FeeReimbursement(maxFee).getReimbursement(actualFee)
+  const priceRatio = PriceRatio(paymentFlow.paymentAmounts())
+  if (priceRatio instanceof Error) return priceRatio
 
+  const feeDifference = FeeReimbursement({
+    prepaidFeeAmount: maxFeeAmounts,
+    priceRatio,
+  }).getReimbursement(actualFeeAmount)
   if (feeDifference instanceof Error) {
-    logger.warn({ maxFee, actualFee }, `Invalid reimbursement fee`)
+    logger.warn(
+      { maxFee: maxFeeAmounts, actualFee: actualFeeAmount },
+      `Invalid reimbursement fee`,
+    )
     return true
   }
 
   // TODO: only reimburse fees is this is above a (configurable) threshold
   // ie: adding an entry for 1 sat fees may not be the best scalability wise for the db
-  if (feeDifference === 0) {
+  if (feeDifference.btc.amount === 0n) {
     return true
   }
 
-  const price = await getCurrentPrice()
-  let amountDisplayCurrency: DisplayCurrencyBaseAmount
-  if (price instanceof Error) {
-    amountDisplayCurrency = toDisplayCurrencyBaseAmount(0)
-  } else {
-    amountDisplayCurrency = DisplayCurrencyConverter(price).fromSats(feeDifference)
+  const {
+    btcPaymentAmount: { amount: satsAmount },
+    usdPaymentAmount: { amount: centsAmount },
+    btcProtocolFee: { amount: satsFee },
+    usdProtocolFee: { amount: centsFee },
+  } = paymentFlow
+
+  const metadata: FeeReimbursementLedgerMetadata = {
```

Comment:
Shouldn't this be returned from a function in the ledger facade? cc @UncleSamtoshi